In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam

2024-10-29 18:20:18.367465: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [2]:
df = pd.read_csv("https://raw.githubusercontent.com/4GeeksAcademy/machine-learning-content/master/assets/clean-pima-indians-diabetes.csv")

df.columns = ['x1', 'x2', 'x3', 'x4', 'x5', 'x6', 'x7', 'x8', 'target']
df.sample(10)

x1        x2        x3        x4        x5        x6        x7  \
402  0.342981  0.472758  0.770014  1.283638  0.071204  0.381700 -0.561368   
574 -0.844885  0.691838  0.873409  0.593630  2.172465 -0.240205  1.268823   
336 -1.141852 -0.121888 -3.572597 -1.288212 -0.692891  0.229397  1.389628   
41   0.936914  0.378867  0.770014 -1.288212 -0.692891  1.041680  0.676880   
43   1.530847  1.568158  2.114155  0.217261  1.391004  1.701660  0.752383   
348 -0.250952 -0.685236 -0.367337 -0.096379 -0.050356 -1.293634 -0.582509   
351  0.046014  0.504055  0.770014 -1.288212 -0.692891 -0.100593 -0.664053   
472 -1.141852 -0.059293 -0.160546  0.405445 -0.692891  0.863993 -0.642912   
736 -1.141852  0.159787  0.873409  0.405445  0.349057 -0.582887  0.130239   
737  1.233880 -1.749339  0.149641  0.154533 -0.692891  0.000942  0.386949   

           x8  target  
402  0.149679       1  
574 -0.871374       0  
336  0.915469       0  
41   0.319855       0  
43   1.766346       1  
348 -0.616111       0  
351 -0.275760       0  
472 -0.956462       0  
736 -1.041549       0  
737  0.745293       0

In [3]:
df_train, df_test = train_test_split(df, test_size=0.2, random_state=2024)

X_train = df_train.drop('target', axis=1).reset_index(drop=True)
y_train = df_train['target'].reset_index(drop=True)
X_test = df_test.drop('target', axis=1).reset_index(drop=True)
y_test = df_test['target'].reset_index(drop=True)

In [4]:
X_train.describe().T

count      mean       std       min       25%       50%       75%  \
x1  614.0 -0.012992  0.992403 -1.141852 -0.844885 -0.250952  0.639947   
x2  614.0  0.015891  1.021084 -3.783654 -0.685236 -0.090591  0.652717   
x3  614.0  0.018123  1.048340 -3.572597 -0.263941  0.149641  0.563223   
x4  614.0  0.011301  0.989202 -1.288212 -1.288212  0.154533  0.766132   
x5  614.0  0.009971  1.004505 -0.692891 -0.692891 -0.397672  0.392471   
x6  614.0  0.012125  1.029201 -4.060474 -0.570195  0.051710  0.584771   
x7  614.0  0.021485  1.041668 -1.168412 -0.706334 -0.322779  0.533425   
x8  614.0 -0.011904  0.997436 -1.041549 -0.786286 -0.360847  0.575118   

         max  
x1  3.906578  
x2  2.444478  
x3  2.734528  
x4  2.663656  
x5  5.767184  
x6  4.455807  
x7  5.883565  
x8  4.063716

In [7]:
model = Sequential([
    Dense(12, input_shape=(X_train.shape[1], ), activation='relu'),
    Dense(8, activation='relu'),
    Dense(1, activation='sigmoid') # When the case is binary you should use 'sigmoid'
])

In [8]:
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_3 (Dense)                 │ (None, 12)             │           108 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 8)              │           104 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 1)              │             9 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 221 (884.00 B)

 Trainable params: 221 (884.00 B)

 Non-trainable params: 0 (0.00 B)

In [9]:
model.fit(X_train, y_train, epochs=10, batch_size=10)

Epoch 1/10
62/62 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.3940 - loss: 0.8119
Epoch 2/10
62/62 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6159 - loss: 0.6668
Epoch 3/10
62/62 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7111 - loss: 0.5905
Epoch 4/10
62/62 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7225 - loss: 0.5469
Epoch 5/10
62/62 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7617 - loss: 0.5055
Epoch 6/10
62/62 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7759 - loss: 0.4945
Epoch 7/10
62/62 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7344 - loss: 0.5284
Epoch 8/10
62/62 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7297 - loss: 0.5034
Epoch 9/10
62/62 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7604 - loss: 0.5138
Epoch 10/10
62/62 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7574 - loss: 0.4661


In [10]:
loss, accuracy = model.evaluate(X_test,  y_test)
print(f"Loss {loss}, Accuracy: {accuracy}")

5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7871 - loss: 0.4392
Loss 0.45574697852134705, Accuracy: 0.7727272510528564


TEST: Accuracy: 0.7727272510528564 TRAIN: Accuracy: 0.7882

There is no big difference => no overfitting